File: c++/94_Binary_Tree_Inorder_Traversal.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// simple iteration solution
class Solution1{
public:
	vector<int> inorderTraversal(TreeNode* root) {
		vector<int> res;
		inOrder(root, res);
        return res;
    }

    void inOrder(TreeNode*Node, vector<int> & res) {
        if (!Node)
            return;
    	//if (Node->val == NULL)
    	//	return;
    	inOrder(Node->left, res);
    	res.push_back(Node->val);
    	inOrder(Node->right, res);
    }
};

/*
   1
  / \
 2   3
    /
   4
    \
     5
*/

class Solution{
public:
	vector<int> inorderTraversal(TreeNode* root) {
		vector<int> res;
		stack<TreeNode*> stk;
		TreeNode *tmp = root;

		while (tmp or ! stk.empty()) {
			if (tmp) {
				stk.push(tmp);				
				tmp = tmp->left;
			}
			else {
				tmp = stk.top();
				stk.pop();
				res.push_back(tmp->val);
				tmp = tmp->right;
			}
		}
		return res;
	}
};


int main() {
	Solution s;
}

File: python/94_Binary_Tree_Inorder_Traversal.py
#!/usr/bin/python
import unittest


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        tmp = root
        res, stk = [], []
        while tmp or stk != []:
        	if tmp:
        		stk.append(tmp)
        		tmp = tmp.left
        	else:
        		tmp = stk[-1]
        		stk.pop()
        		res.push_back(tmp.val)
        		tmp = tmp.right
        return res


class TestSolution(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test_method(self):
        """Such as self.assertEqual, self.assertTrue"""
        pass


if __name__ == "__main__":
    unittest.main()